File: ocaml/fenwick_tree.ml
class fenwick_tree n zero op =
  object (self: 'self)
  val a = Array.make n zero
  method update x (v:int) =
    let rec loop i =
      if i < n then begin
        a.(i) <- op a.(i) v;
        loop (i lor (i + 1))
      end
    in
    loop x
  method reduce x =
    let rec loop r i = if i >= 0 then loop (op r a.(i)) ((i land (i + 1)) - 1) else r in
    loop zero x
end;;

File: c++/FenwickTree.cpp
#include <vector>
using namespace std;
template<typename T> class FenwickTree {
  vector<T> a;
  public:
  FenwickTree (const int n) : a (n, (T) 0) {}
  void update (const int x, const T val) {
    const int n = (int) a.size ();
    for (int i = x; i < n; i |= i + 1) {
      a[i] += val;
    }
  }
  T reduce (const int x) const {
    T r = 0;
    for (int i = x; i >= 0; i = (i & (i + 1)) - 1) {
      r += a[i];
    }
    return r;
  }
};
